src/main/java/teammates/storage/search/StudentSearchDocument.java

```diff
@@ -0,0 +1,88 @@
+package teammates.storage.search;
+
+import java.util.Comparator;
+import java.util.List;
+
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.common.SolrDocument;
+import org.apache.solr.common.SolrInputDocument;
+
+import teammates.common.datatransfer.StudentSearchResultBundle;
+import teammates.common.datatransfer.attributes.CourseAttributes;
+import teammates.common.datatransfer.attributes.InstructorAttributes;
+import teammates.common.datatransfer.attributes.StudentAttributes;
+import teammates.common.util.StringHelper;
+
+/**
+ * The {@link SearchDocument} object that defines how we store {@link SolrInputDocument} for students.
+ */
+class StudentSearchDocument extends SearchDocument {
+
+    private final StudentAttributes student;
+
+    StudentSearchDocument(StudentAttributes student) {
+        this.student = student;
+    }
+
+    @Override
+    SolrInputDocument toDocument() {
+        SolrInputDocument document = new SolrInputDocument();
+
+        CourseAttributes course = coursesDb.getCourse(student.course);
+
+        document.addField("id", student.key);
+        document.addField("name", student.getName());
+        document.addField("email", student.getEmail());
+        document.addField("courseId", student.getCourse());
+        document.addField("courseName", course == null ? "" : course.getName());
+        document.addField("team", student.getTeam());
+        document.addField("section", student.getSection());
+
+        return document;
+    }
+
+    /**
+     * Produces a {@link StudentSearchResultBundle} from the {@code QueryResponse} collection.
+     *
+     * <p>The list of {@link InstructorAttributes} is used to filter out the search result.</p>
+     */
+    static StudentSearchResultBundle fromResponse(QueryResponse response,
+                                                        List<InstructorAttributes> instructors) {
+        if (response == null) {
+            return new StudentSearchResultBundle();
+        }
+
+        List<SolrDocument> filteredResults = filterOutCourseId(response, instructors);
```

Comment:
This means that the initial query is performed against the entire DB, then filtered out by its course ID here.

Now suppose that there are (X + 1) Alices in the DB, where X is the number of the returned results, and the Alice that the searcher is interested in is the (X + 1)-th result. How can we make sure that it will be returned? The others are of no usage to the searcher (and will be filtered out anyway).

Reply:
Added in filterQuery, now only relevant results are fetched. No more local filter.